File: src/main/native/cpp/CKSimAdvancedObject.cpp
#include "team195/CKSimAdvancedObject.h"
#include "frc/smartdashboard/SendableBuilder.h"
#include "frc/smartdashboard/SendableRegistry.h"
#include <iostream>

#include "team195/CKSim.h"

namespace team195
{
    CKSimAdvancedObject::CKSimAdvancedObject(int id) : id(id)
    {
        if (!validateId(id))
        {
            std::cout << "Invalid advanced object ID: " << id << std::endl;
            id = -1;
            return;
        }
        if (!CKSim::GetInstance().registerAdvObj(id))
        {
            std::cout << "Failed to add advanced object ID: " << id << "; ID Already in use!" << std::endl;
            id = -1;
            return;
        }
        frc::SendableRegistry::GetInstance().AddLW(this, "CKSimAdvanced", id);
    }

    float CKSimAdvancedObject::GetValue()
    {
        return CKSim::GetInstance().getAdvObjValue(id);
    }

    bool CKSimAdvancedObject::validateId(int id)
    {
        return (id >= 0 && id < CKSim::MAX_NUM_ADVOBJ);
    }

    void CKSimAdvancedObject::InitSendable(frc::SendableBuilder &builder)
    {
        builder.SetSmartDashboardType("CKSimAdvanced");
        // builder.SetActuator(true);
        // builder.SetSafeState([=]() { SetDisabled(); });
        builder.AddDoubleProperty(
            "Value", [=]() { return (double)GetValue(); }, [=](double value) { NULL; });
    }
} // namespace team195

File: src/main/java/com/team195/CKSimAdvancedObject.java
package com.team195;

import edu.wpi.first.wpilibj.Sendable;
import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;
import edu.wpi.first.wpilibj.smartdashboard.SendableRegistry;

public class CKSimAdvancedObject implements Sendable {
    private int id;

    public CKSimAdvancedObject(int id) {
        if (!validateId(id)) {
            System.out.println("Invalid advanced object ID: " + id);
            return;
        }
        if (!CKSim.getInstance().registerAdvObj(id)) {
            System.out.println("Failed to add advanced object ID: " + id + "; ID Already in use!");
            return;
        }
        this.id = id;
        SendableRegistry.addLW(this, "CKSimAdvanced", id);
    }

    public float getValue() {
        return CKSim.getInstance().getAdvObjValue(id);
    }

    private boolean validateId(int id) {
        return (id >= 0 && id < CKSim.MAX_NUM_ADVOBJ);
    }

    @Override
    public void initSendable(SendableBuilder builder) {
        builder.setSmartDashboardType("CKSimAdvanced");
        // builder.setActuator(false);
        // builder.setSafeState(this::setDisabled);
        builder.addDoubleProperty("Value", this::getValue, null);
    }
}
